W-15278074 Adding Java 17 support for Studio (#230)
mulesoft/docs-general: modules/ROOT/pages/java-support.adoc

[IMPORTANT]
To ensure your API proxies or Mule apps are protected when upgrading, upgrade your API policies _before_ upgrading your API proxies or Mule apps.

[[application-performance]]
== Java 17 Performance Considerations

Java 17 provides important performance changes to MuleSoft products. Java 17 includes a new JIT compiler, which improves code execution speed. 

To enhance application performance, see:

* <<memory-management>>
* <<default-tls-cipher-selection>>

[[memory-management]]
=== Memory Management

Java 17 manages memory differently than it did in earlier versions like Java 8. Java 17 reduces request latency and executes work more predictably. To achieve this, it adopts a different memory management approach. Although highly dependent on each application, in some cases, Java 17 preferentially commits to heap memory.

Because of this change in Java 17 memory management, to ensure optimal operation and effectively leverage these improvements, you must conduct xref:mule-runtime::tuning-test-validations.adoc[performance testing] and xref:mule-runtime::tuning-monitoring.adoc[monitoring] for your most common workloads.

If your apps are deployed on-premises:

> * If your workloads are running on Runtime Fabric, you must allocate 1400 MB of memory for 0.50 vCPU replica sizes. Memory requirements vary for individual apps. This is the minimum recommended size as the best configuration for memory allocation for most applications.

Reviewer: ```suggestion
* If your workloads run on Runtime Fabric, you must allocate at least 1400 MB of memory for 0.50 vCPU replica sizes. Because memory requirements vary for individual apps, this minimum provides the best memory allocation configuration for most applications.
```

Reviewer: (Trying to avoid the orphaned "This".)

> * If your workloads are running on-premises, such as in a hybrid or standalone environment, you must tune your operating system’s garbage collection algorithm. 

Reviewer: ```suggestion
* If your workloads run on-premises, such as in a hybrid or standalone environment, you must tune your operating system’s garbage collection algorithm. 
```

[[default-tls-cipher-selection]]
=== Default TLS Cipher Selection

When using TLS in a server with Java 11 and later, the server determines the cipher preference. The client supported ciphers list is used to filter, not to determine priority. See https://bugs.openjdk.org/browse/JDK-8168261[Use server cipher suites preference by default (JDK-8168261)].

> Upgrading your server from Java 8 to Java 11 and later can cause negotiation of a different cipher (depending on the supported ciphers list of the client), which can affect the trade-off between security and performance.

Reviewer: cause cipher renegotiation?

> If you want to change the cipher's default values, see xref:mule-runtime::tls-configuration.adoc#optional-specify-protocols-and-cipher-suites[control the cipher selection].

Reviewer: Since this is a "see", suggest using the actual heading in the link text.

. In Anypoint Studio, update the connectors in your app to the latest Java 17-compatible version. 
. xref:mule-runtime::deploying.adoc[Deploy your applications] to a target running on Mule 4.6 or later.

[[anypoint-studio]]
== Anypoint Studio

> Starting with Anypoint Studio (Studio) 7.17, if your Mule app is running on Mule runtime 4.6 or later, you can change the target JVM at the Studio project level to upgrade or downgrade the JVM. When you compile your app, Studio 7.17 offers real-time guidance and alerts you if there are any incompatible connector versions or project mismatches to prevent deployment failures.

Reviewer: These 4 sentences that start with "When" seem like bullet points, which might be more scannable?

When you add a new connector or module to your project, Studio shows the supported JVM version for each version of the connector or module. 

When you upgrade an existing Studio project to Java 17, Studio automatically searches Exchange and suggests which extensions (modules and connectors) in your app to upgrade. 

> When you deploy an app to CloudHub from Studio, Runtime Manager proactively detects mismatches between the project's Java version and the deployment environment. For example, if your project is built for Java 8 and your target environment is a Java 17 environment, Runtime Manager and Studio provides guidance to prevent deployment failures.

Reviewer: ```suggestion
When you deploy an app to CloudHub from Studio, Runtime Manager proactively detects mismatches between the project's Java version and the deployment environment. For example, if your project is built for Java 8 and your target environment is Java 17, Runtime Manager and Studio provides guidance to prevent deployment failures.
```

Reviewer: (to avoid the environment...environment)

* Verify that the objects used by your application are Plain Old Java Objects (POJOs). 
+

> POJOs are required from DataWeave 6.0.0 on, and you must ensure that the POJOs have:

Reviewer: I'm not sure about the "from xxx on". Maybe "for xxx and later"?
